PARQUET-2227: Refactor several file rewriters to use a new unified ParquetRewriter implementation

wgtmac opened the pull request:
### Jira

- This patch aims to solve the first step of [PARQUET-2075](https://issues.apache.org/jira/browse/PARQUET-2075).

### Tests

- Make sure all tasks pass, especially CompressionConverterTest, ColumnPrunerTest, ColumnMaskerTest, and ColumnEncryptorTest.

### Commits

 - A new ParquetRewriter is introduced to unify rewriting logic.
 - RewriteOptions is defined to provide essential settings.
 - CompressionConverter, ColumnPruner, ColumnMasker, and ColumnEncryptor have been refactored and marked as deprecated.

wgtmac commented:
Can you please take a look when you have time? @shangxinli @gszadovszky @ggershinsky

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/RewriteOptions.java:
```diff
@@ -0,0 +1,144 @@
+                         Path inputFile,
+                         Path outputFile,
+                         List<String> pruneColumns,
+                         CompressionCodecName codecName,
```
Maybe use the variable 'newCodecName'?

wgtmac replied:
Fixed

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/RewriteOptions.java:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    public RewriteOptions build() {
+      // TODO: validate any conflict setting
```
Can you open a Jira to track it? If you can do it in this PR, it is even better.

wgtmac replied:
Fixed and removed the TODO

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+    // Get file metadata and full schema from the input file
+    meta = ParquetFileReader.readFooter(conf, inPath, NO_FILTER);
+    schema = meta.getFileMetaData().getSchema();
+    createdBy = meta.getFileMetaData().getCreatedBy();
```
There was some discussion should we carry over the old author or replace it with this rewriter author. Or we can append 'reCreatedBy' but it needs a specification (parquet-format) change. Any thoughts on it?

wgtmac replied:
IMHO, reusing the old `createdBy` is a bad idea which makes it difficult to reason about the origin of the file. What about concatenating the old `createdBy` with the writer version of `ParquetRewriter` so we can keep both old and new creator in the same field?

wgtmac replied:
I'd rather keep the same behavior for now and fix it in a separate JIRA.

gszadovszky replied:
Since we have only one field for _created by_ and it's content is more or less specified, we should write the current version of parquet-mr there. It is also a good idea to keep the original _created by_ value, though. What do you think about adding it to the `key_value_metadata` with a specific key? Even though this field would not be specified at least we won't lose the info.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+    // Prune columns if specified
+    if (pruneColumns != null && !pruneColumns.isEmpty()) {
+      List<String> paths = new ArrayList<>();
+      getPaths(schema, paths, null);
```
Should we move the line 120 to be inside the getPath()?

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+      BlockMetaData blockMetaData = meta.getBlocks().get(blockId);
+      List<ColumnChunkMetaData> columnsInOrder = blockMetaData.getColumns();
+
+      for (int i = 0, columnId = 0; i < columnsInOrder.size(); i += 1) {
```
Can we use i++?

wgtmac replied:
Fixed

shangxinli commented:
I just left some comments initially. I will spend more time on it. 

@ggershinsky If you have time, can you have a look too?

shangxinli commented:
If you can add more unit tests, particularly the combinations of prune, mask, trans-compression etc.

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+        if (this.encryptMode) {
+          encryptorRunTime = new EncryptorRunTime(writer.getEncryptor(), chunk, blockId, columnId);
+        }
+        boolean encryptColumn = encryptColumns != null && encryptColumns.contains(chunk.getPath());
```
Can we apply this.encryptMode to this check also?

wgtmac replied:
Fixed

shangxinli commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+      }
+    }
+
+    currentPath.remove(fieldName);
```
This line seems a bug(even in the original code). We should remove by index instead of by name. Something like this: currentPath.remove(currentPath.size() - 1)

wgtmac replied:
Fixed

ggershinsky commented:
> I just left some comments initially. I will spend more time on it.
> 
> @ggershinsky If you have time, can you have a look too?

sure, will do

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+    @Override public GroupConverter asGroupConverter() { return new DummyGroupConverter(); }
+  }
+
+  private static class EncryptorRunTime {
```
This is a per-column-chunk collection, maybe the class name should reflect that?

wgtmac replied:
Renamed it to ColumnChunkEncryptorRunTime

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+                            ColumnChunkMetaData chunk,
+                            int blockId,
+                            int columnId) throws IOException  {
+      if (fileEncryptor == null) {
```
should this object be created at all if the file encryptor is null?

wgtmac replied:
This is a good question. The code was just moved from https://github.com/apache/parquet-mr/blob/1adc22804a700d78f8480667d083e91d6147339f/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java#L84. I think it should throw if the EncryptorRuntime is required but the fileEncryptor is null.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+                             ModuleType moduleType,
+                             int blockId,
+                             int columnId) {
+      if (colEncrSetup != null && colEncrSetup.isEncrypted()) {
```
should this be checked here, upon every call? also, what's the point of creating an EncryptorRunTime if the column is not encrypted?

wgtmac replied:
If `EncryptorRunTime` was not created for non-encrypted columns, generating the file footer will get some error. It can be reproduced by the ColumnEncryptorTest.

ggershinsky replied:
not sureI understand. The EncryptorRunTime object is created per column, so no need to run this check for every module aad; this boolean can be built in the EncryptorRunTime constructor. Also, it would be more logical to use EncryptorRunTime objects only for encrypted columns.

wgtmac replied:
Fixed it by checking `colEncrSetup` only once per creation.

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/ParquetRewriter.java:
```diff
@@ -0,0 +1,733 @@
+    if (options.getEncryptColumns() != null && options.getFileEncryptionProperties() != null) {
+      this.encryptColumns = convertToColumnPaths(options.getEncryptColumns());
+      this.encryptMode = true;
+      // TODO: make sure options.getFileEncryptionProperties() is set
```
what is left to do here?

wgtmac replied:
I forgot to remove this TODO.

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/MaskMode.java:
```diff
@@ -0,0 +1,45 @@
+  private String mode;
+
+  MaskMode(String text) {
+    this.mode = text;
```
Fixed

wgtmac commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/MaskMode.java:
```diff
@@ -0,0 +1,45 @@
+    return this.mode;
+  }
+
+  public static MaskMode fromString(String mode) {
```
Fixed

wgtmac commented:
> If you can add more unit tests, particularly the combinations of prune, mask, trans-compression etc, it would be better.

I have added some test cases in the `ParquetRewriterTest` to mix config of prune, nullify and trans-compression. 

There are two outstanding issues as below:
- Masking and encrypting same column is not supported yet.
- Fix `createdBy` of rewritten file to keep both original info and new rewriter info.

I am inclined to fix them in separate JIRAs because this patch is simply a refactoring patch to unify different rewriters without changing any behavior, and it is large enough.

@ggershinsky @shangxinli Please review again when you have time. Thanks!

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/RewriteOptions.java:
```diff
@@ -0,0 +1,178 @@
+        if (encryptColumns != null) {
+          for (String pruneColumn : pruneColumns) {
+            Preconditions.checkArgument(!encryptColumns.contains(pruneColumn),
+                    "Cannot prune and mask same column");
```
Cannot prune and encrypt same column?

wgtmac replied:
Yes, pruned column does not exist in the rewritten file. Then it does not make sense to encrypt the missing column any more.

ggershinsky replied:
Yep. I meant the exception text in the line 149, it is identical to the masking check (line 142); I guess the encryption check would print "Cannot prune and encrypt same column"

wgtmac replied:
Sorry, I misunderstood your meaning. Fixed the error message. Please take a look again. Thanks!

ggershinsky commented on parquet-hadoop/src/main/java/org/apache/parquet/hadoop/rewrite/RewriteOptions.java:
```diff
@@ -0,0 +1,178 @@
+      }
+
+      if (encryptColumns != null && !encryptColumns.isEmpty()) {
+        Preconditions.checkArgument(fileEncryptionProperties != null,
```
what if `fileEncryptionProperties` was set, but the `encryptColumns` is null or `encryptColumns.isEmpty()`?

wgtmac replied:
I have added checking to disable this case.

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/hadoop/rewrite/ParquetRewriterTest.java:
```diff
@@ -0,0 +1,308 @@
+import org.junit.Test;
+
+import java.io.IOException;
+import java.util.*;
```
please don't use * (also in lines 57, 58)

wgtmac replied:
Fixed

ggershinsky commented on parquet-hadoop/src/test/java/org/apache/parquet/hadoop/rewrite/ParquetRewriterTest.java:
```diff
@@ -0,0 +1,308 @@
+  private ParquetRewriter rewriter = null;
+
+  @Test
+  public void testPruneSingleColumnAndTranslateCodec() throws Exception {
```
do we also have an encryption rewrite unitest?

wgtmac replied:
Added a test case to prune, transcode and encrypt columns.

shangxinli commented:
@gszadovszky I Just want to check if you have time to have a look. @wgtmac Just be nice to take over the work that we discussed earlier to have an aggregated rewriter.

gszadovszky commented:
> @gszadovszky I Just want to check if you have time to have a look. @wgtmac just be nice to take over the work that we discussed earlier to have an aggregated rewriter.

@shangxinli, I'll try to take a look this week.

shangxinli commented:
Thanks a lot @gszadovszky

gszadovszky reviewed (changes requested):
I think it is a great refactor. Thanks a lot for working on it, @wgtmac!
In the other hand I've thought about PARQUET-2075 as a request for a new feature in `parquet-cli` that can be used to convert from one parquet file to another with specific configurations. (Later on we might extend it to allow multiple parquet files to be merged/rewritten to one specified and the tool would decide which level of deserialization/serialization is required.)
I am fine with handling it in a separate jira but let's make it clear. Either create another jira for this refactor as a prerequisite of PARQUET-2075 or rephrase PARQUET-2075 and create a new for `parquet-cli`.
@shangxinli, what do you think?

wgtmac commented:
> I think it is a great refactor. Thanks a lot for working on it, @wgtmac! In the other hand I've thought about PARQUET-2075 as a request for a new feature in `parquet-cli` that can be used to convert from one parquet file to another with specific configurations. (Later on we might extend it to allow multiple parquet files to be merged/rewritten to one specified and the tool would decide which level of deserialization/serialization is required.) I am fine with handling it in a separate jira but let's make it clear. Either create another jira for this refactor as a prerequisite of PARQUET-2075 or rephrase PARQUET-2075 and create a new for `parquet-cli`. @shangxinli, what do you think?

Thanks for your review @gszadovszky 

- I'd prefer creating a new JIRA for this refactor to be a prerequisite. Merging multiple files to a single one with customized pruning, encryption, and codec is also in my mind and will be supported later. I will create separate JIRAs as sub-tasks of PARQUET-2075 and work on them progressively.
- Putting the original `created_by` into `key_value_metadata` is a good idea. However, it is tricky if a file has been rewritten for several times. What about adding a key named `original_created_by` to `key_value_metadata` and concatenating all old `created_by`s to it?

gszadovszky commented:
> * I'd prefer creating a new JIRA for this refactor to be a prerequisite. Merging multiple files to a single one with customized pruning, encryption, and codec is also in my mind and will be supported later. I will create separate JIRAs as sub-tasks of PARQUET-2075 and work on them progressively.

Perfect! :)

> * Putting the original `created_by` into `key_value_metadata` is a good idea. However, it is tricky if a file has been rewritten for several times. What about adding a key named `original_created_by` to `key_value_metadata` and concatenating all old `created_by`s to it?

It sounds good to me. Maybe have the latest one at the beginning and use the separator `'\n'`?

wgtmac commented:
> > * I'd prefer creating a new JIRA for this refactor to be a prerequisite. Merging multiple files to a single one with customized pruning, encryption, and codec is also in my mind and will be supported later. I will create separate JIRAs as sub-tasks of PARQUET-2075 and work on them progressively.
> 
> Perfect! :)
> 
> > * Putting the original `created_by` into `key_value_metadata` is a good idea. However, it is tricky if a file has been rewritten for several times. What about adding a key named `original_created_by` to `key_value_metadata` and concatenating all old `created_by`s to it?
> 
> It sounds good to me. Maybe have the latest one at the beginning and use the separator `'\n'`?

I am afraid some implementations may drop characters after `'\n'` when displaying the string content. Let me do some investigation.

gszadovszky commented:
> I am afraid some implementations may drop characters after `'\n'` when displaying the string content. Let me do some investigation.

I do not have a strong opinion for `'\n'` only that we need a character that probably won't be used by any systems writing parquet files.

wgtmac commented:
> > I am afraid some implementations may drop characters after `'\n'` when displaying the string content. Let me do some investigation.
> 
> I do not have a strong opinion for `'\n'` only that we need a character that probably won't be used by any systems writing parquet files.

As we are discussing a new entry (`original.created.by`) to the key value metadata, I need to raise two related issues once we have supported rewriting (merging) several files into one:
- We need to merge `original.created.by` from all input files, making it difficult to tell which created_by comes from which input file. Therefore, `original.created.by` should be dropped in this case.
- Is there any key value metadata that will conflict from different input files and should be dealt with by the rewriter? For now we simply keep all the old key value metadata from the old file.

@gszadovszky @ggershinsky @shangxinli Thoughts?

If this behavior requires further discussion, I'd suggest to keep the current state of `created_by` unchanged in this pull request which is large enough. All rewriters (ColumnPruner, CompressionConverter, ColumnMasker, and ColumnEncrypter) have dropped original `created_by` and store the current writer version to the footer.

gszadovszky commented:
I agree that merging the key-value metadata is not an easy question. Let's discuss it separately as it is not related to this PR.

I also agree to store the current writer (parquet-mr) in `created_by` in case of rewriting.  It is not easy to decide what would be the proper solution anyway. `created_by` is usually used for handling potential erroneous writes. Let's say there was an issue in parquet-mr at the version 1.2.3 that written a specific encoding of integers wrongly (not according to spec). What if we rewrite the file but do not re-encode the pages? Can we still handle the original issue? What if the rewriter re-encodes the related pages?
Let's store the original writer in `original.created.by` for now. Let's discuss this topic separately however, I am not sure if we can find a proper solution.

wgtmac commented:
> I agree that merging the key-value metadata is not an easy question. Let's discuss it separately as it is not related to this PR.
> 
> I also agree to store the current writer (parquet-mr) in `created_by` in case of rewriting. It is not easy to decide what would be the proper solution anyway. `created_by` is usually used for handling potential erroneous writes. Let's say there was an issue in parquet-mr at the version 1.2.3 that written a specific encoding of integers wrongly (not according to spec). What if we rewrite the file but do not re-encode the pages? Can we still handle the original issue? What if the rewriter re-encodes the related pages? Let's store the original writer in `original.created.by` for now. Let's discuss this topic separately however, I am not sure if we can find a proper solution.

I agree. Now I have updated this PR to preserve the old writer version into `original.created.by` and added a test to make sure it is preserved. Please take a look. Thanks!

wgtmac commented:
Gentle ping. @gszadovszky @ggershinsky @shangxinli

Any chance to take another look?

gszadovszky commented:
Thank you, @wgtmac for working on this! It looks good to me.
Since there were other reviewers, I would wait a bit to give a chance to them for add feedback. I'll push it after a couple of days. (Feel free to ping me if I forget.)

shangxinli commented:
Thanks @wgtmac for working on this and Thanks @gszadovszky and @ggershinsky for reviewing it! I am a little late for the comments discussion but I see we are in the right direction. Let's address it in a separate discussion. If it turns out that changing the parquet-format is the right way to solve it, we can make the proposal and I can help for the approval. 

My comments are all addressed. I don't have further comments.

ggershinsky reviewed (approved):
My comments are addressed too. Thanks for working on this PR.

wgtmac commented:
Thank you @gszadovszky @ggershinsky @shangxinli
